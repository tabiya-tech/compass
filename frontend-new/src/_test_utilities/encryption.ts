import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
