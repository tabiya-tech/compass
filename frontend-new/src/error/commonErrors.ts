enum CommonError {
  TOKEN_ERROR = "TokenError",
  AUTHENTICATION_ERROR = "AuthenticationError",
  USER_PREFERENCE_ERROR = "UserPreferenceError",
  INVITATION_ERROR = "InvitationError",
  SESSION_ERROR = "SessionError",
  CHAT_ERROR = "ChatError",
  REACTION_ERROR = "ReactionError",
  FEEDBACK_ERROR = "FeedbackError",
  ENV_ERROR = "EnvError",
  COMPONENT_ERROR = "ComponentError",
  CONFIGURATION_ERROR = "ConfigurationError",
}

/**
 * Used for errors related to token generation, decoding and verification
 * */
export class TokenError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.TOKEN_ERROR;
    this.cause = cause;
  }
}

export class AuthenticationError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.AUTHENTICATION_ERROR;
    this.cause = cause;
  }
}

export class UserPreferenceError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.USER_PREFERENCE_ERROR;
    this.cause = cause;
  }
}

export class SessionError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.SESSION_ERROR;
    this.cause = cause;
  }
}

export class ChatError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.CHAT_ERROR;
    this.cause = cause;
  }
}

export class ReactionError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.REACTION_ERROR;
    this.cause = cause;
  }
}

export class FeedbackError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.FEEDBACK_ERROR;
    this.cause = cause;
  }
}

export class EnvError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.ENV_ERROR;
    this.cause = cause;
  }
}

export class ComponentError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.COMPONENT_ERROR;
    this.cause = cause;
  }
}

export class InvitationError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.INVITATION_ERROR;
    this.cause = cause;
  }
}

export class ConfigurationError extends Error {
  constructor(message: string, cause?: Error | string) {
    super(message);
    this.name = CommonError.CONFIGURATION_ERROR;
    this.cause = cause;
  }
}
