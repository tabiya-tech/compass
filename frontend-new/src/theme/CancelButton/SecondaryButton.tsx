import { Button, ButtonProps } from "@mui/material";
import React, { useContext } from "react";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";

interface SecondaryButtonProps extends ButtonProps {
  // Add additional props specific to SecondaryButton Button here
  disableWhenOffline?: boolean;
}

const SecondaryButton: React.FC<SecondaryButtonProps> = ({
  style,
  disabled,
  children,
  disableWhenOffline,
  ...props
}: Readonly<SecondaryButtonProps>) => {
  const isOnline = useContext(IsOnlineContext);

  return (
    // props are passed to the component last, so that they can override the default values
    <Button
      variant={"text"}
      style={style}
      sx={{
        borderRadius: (theme) => theme.tabiyaRounding.xl,
        paddingY: (theme) => theme.fixedSpacing(theme.tabiyaSpacing.xs),
        paddingX: (theme) => theme.fixedSpacing(theme.tabiyaSpacing.md),
        color: (theme) => theme.palette.text.secondary
      }}
      disableElevation
      disabled={Boolean(disabled || (disableWhenOffline && !isOnline))}
      {...props}
    >
      {children ?? "Click here"}
    </Button>
  );
};

export default SecondaryButton;
