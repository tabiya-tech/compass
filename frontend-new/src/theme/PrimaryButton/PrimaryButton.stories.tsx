import type { Meta, StoryObj } from "@storybook/react";

import PrimaryButton from "./PrimaryButton";

const meta: Meta<typeof PrimaryButton> = {
  title: "Components/PrimaryButton",
  component: PrimaryButton,
  tags: ["autodocs"],
  argTypes: {
    onClick: { action: "clicked" },
  },
};

export default meta;
type Story = StoryObj<typeof PrimaryButton>;

export const Shown: Story = {
  args: {
    children: "Click here",
  },
};

export const Disabled: Story = {
  args: {
    children: "Click here",
    disabled: true,
  },
};

export const DisabledWhenOffline: Story = {
  args: {
    children: "Click here",
    disableWhenOffline: true,
  },
};
