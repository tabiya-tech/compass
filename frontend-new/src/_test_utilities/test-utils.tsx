// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
