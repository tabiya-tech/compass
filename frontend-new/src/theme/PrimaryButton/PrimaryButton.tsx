import { Button, ButtonProps } from "@mui/material";
import React, { useContext } from "react";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";
import { ComponentError } from "src/error/commonErrors";

interface PrimaryButtonProps extends ButtonProps {
  // Add additional props specific to PrimaryButton Button here
  disableWhenOffline?: boolean;
}

const PrimaryButton: React.FC<PrimaryButtonProps> = ({
  style,
  disabled,
  children,
  disableWhenOffline,
  sx,
  ...props
}: Readonly<PrimaryButtonProps>) => {
  const isOnline = useContext(IsOnlineContext);

  if (!children) {
    throw new ComponentError("Children are required for PrimaryButton component");
  }

  return (
    // props are passed to the component last, so that they can override the default values
    <Button
      variant={"contained"}
      color={"primary"}
      style={style}
      sx={{
        borderRadius: (theme) => theme.fixedSpacing(theme.tabiyaRounding.xl),
        paddingY: (theme) => theme.fixedSpacing(theme.tabiyaSpacing.xs),
        paddingX: (theme) => theme.fixedSpacing(theme.tabiyaSpacing.md),
        ...sx,
      }}
      disableElevation
      disabled={Boolean(disabled || (disableWhenOffline && !isOnline))}
      {...props}
    >
      {children ?? "Click here"}
    </Button>
  );
};

export default PrimaryButton;
