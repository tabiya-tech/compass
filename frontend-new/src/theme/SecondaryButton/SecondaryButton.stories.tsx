import type { Meta, StoryObj } from "@storybook/react";

import SecondaryButton from "./SecondaryButton";

const meta: Meta<typeof SecondaryButton> = {
  title: "Components/SecondaryButton",
  component: SecondaryButton,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof SecondaryButton>;

export const Shown: Story = {
  args: {
    children: "Click here",
  },
};

export const DisabledWhenOffline: Story = {
  args: {
    children: "Click here",
    disableWhenOffline: true,
  },
};
