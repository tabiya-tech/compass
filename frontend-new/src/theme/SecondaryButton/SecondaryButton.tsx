import { ButtonProps } from "@mui/material";
import React, { useContext } from "react";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";
import { ComponentError } from "src/error/commonErrors";

interface SecondaryButtonProps extends ButtonProps {
  // Add additional props specific to SecondaryButton Button here
  disableWhenOffline?: boolean;
}

const SecondaryButton: React.FC<SecondaryButtonProps> = ({
  style,
  disabled,
  children,
  disableWhenOffline,
  sx,
  ...props
}: Readonly<SecondaryButtonProps>) => {
  const isOnline = useContext(IsOnlineContext);

  if(!children) {
    throw new ComponentError("Children are required for SecondaryButton component");
  }

  return (
    // props are passed to the component last, so that they can override the default values
    <PrimaryButton
      variant={"text"}
      style={style}
      sx={{
        color: (theme) => theme.palette.text.secondary,
        ...sx
      }}
      disabled={Boolean(disabled || (disableWhenOffline && !isOnline))}
      {...props}
    >
      {children ?? "Click here"}
    </PrimaryButton>
  );
};

export default SecondaryButton;
