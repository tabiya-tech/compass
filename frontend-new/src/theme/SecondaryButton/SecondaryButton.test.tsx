// mute the console
import "src/_test_utilities/consoleMock";

import { render, screen } from "src/_test_utilities/test-utils";
import SecondaryButton from "./SecondaryButton";
import { unmockBrowserIsOnLine } from "src/_test_utilities/mockBrowserIsOnline";
import { ComponentError } from "src/error/commonErrors";
import PrimaryButton from "src/theme/PrimaryButton/PrimaryButton";

// mock the primary button component
jest.mock("src/theme/PrimaryButton/PrimaryButton", () => {
  return jest.fn().mockImplementation((props) => <div data-testid={props["data-testid"]}> {props.children}</div>);
});

describe("Secondary Button tests", () => {
  beforeEach(() => {
    unmockBrowserIsOnLine();
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  test("should render the primary button with custom props", () => {
    // GIVEN a data test id
    const givenTestId = "foo";
    const givenText = "Bar";
    // WHEN the component is rendered
    render(<SecondaryButton data-testid={givenTestId}>{givenText}</SecondaryButton>);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND  the component should call the primary button component with the correct data test id
    const actualPrimaryButton = screen.getByTestId(givenTestId);
    expect(actualPrimaryButton).toBeInTheDocument();

    // AND the component should be called with the correct props
    expect(PrimaryButton).toHaveBeenCalledWith(
      expect.objectContaining({
        variant: "text",
        style: undefined,
        sx: expect.objectContaining({
          color: expect.any(Function),
        }),
        disabled: false,
        children: givenText,
      }),
      {}
    );

    // AND the component should be called with the correct children
    expect(actualPrimaryButton).toHaveTextContent(givenText);

    // AND the component should match the snapshot
    expect(actualPrimaryButton).toMatchSnapshot();
  });

  test("should throw an error when no children are provided", () => {
    // GIVEN no children
    // WHEN the component is rendered
    // THEN expect an error to have occurred
    expect(() => render(<SecondaryButton />)).toThrow(
      new ComponentError("Children are required for SecondaryButton component")
    );
  });
});
