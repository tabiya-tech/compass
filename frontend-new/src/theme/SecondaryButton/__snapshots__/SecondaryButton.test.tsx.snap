// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secondary Button tests should render the primary button with custom props 1`] = `
<div
  data-testid="foo"
>
   
  Bar
</div>
`;
